src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+#include <IO/ReadHelpers.h>
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <type_traits>
+#include <unordered_map>
+#include <Processors/Formats/Impl/NpyRowInputFormat.h>
+#include <Formats/FormatFactory.h>
+#include <Formats/EscapingRuleUtils.h>
+#include <DataTypes/Serializations/SerializationNullable.h>
+#include <DataTypes/DataTypeString.h>
+#include <Common/typeid_cast.h>
+#include <Common/Exception.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <Storages/IStorage.h>
+#include <Columns/IColumn.h>
+#include <Core/Field.h>
+#include <Core/NamesAndTypes.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/Serializations/ISerialization.h>
+#include <IO/ReadBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <Processors/Formats/IRowInputFormat.h>
+#include <base/types.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+}
+
+
+DataTypePtr createDataType(size_t depth, DataTypePtr nested_type)
```

Comment:
It's better to place functions that are used inly in this cpp file in an anonymous namespace, so they are not visible in other cpp files.
Like this:
```
namespace
{

DataTypePtr createDataType(size_t depth, DataTypePtr nested_type)
{
     ...
}

}
```
And this namespace can have all such functions.

Also consider better naming for this function, maybe smth like `createNestedArrayType`